src/solver.jl

```diff
@@ -9,61 +10,51 @@
 
 
 # Policy struct created according to one used in DiscreteValueIteration
-mutable struct FiniteHorizonValuePolicy{Q<:AbstractArray, U<:AbstractMatrix, P<:AbstractMatrix, A, M<:MDP} <: Policy
+mutable struct FiniteHorizonValuePolicy{Q<:AbstractArray, U<:AbstractMatrix, P<:AbstractMatrix, A, W<:FHWrapper} <: Policy
     qmat::Q
     util::U 
     policy::P 
     action_map::Vector{A}
     include_Q::Bool 
-    mdp::M
+    w::W
 end
 
 # Policy constructor
-function FiniteHorizonValuePolicy(mdp::MDP)
-    return FiniteHorizonValuePolicy(zeros(mdp.horizon, mdp.no_states, length(mdp.actions)), zeros(mdp.horizon, mdp.no_states), ones(Int64, mdp.horizon, mdp.no_states), ordered_actions(mdp), true, mdp)
+function FiniteHorizonValuePolicy(w::FHWrapper)
+    return FiniteHorizonValuePolicy(zeros(w.horizon + 1, length(states(w.m)), length(actions(w.m))), zeros(w.horizon + 1, length(states(w.m))), ones(Int64, w.horizon + 1, length(states(w.m))), ordered_actions(w.m), true, w)
 end
 
 function action(policy::FiniteHorizonValuePolicy, s::S) where S
-    sidx = stage_stateindex(policy.mdp, s, s.epoch)
+    sidx = stateindex(policy.w, s)
     aidx = policy.policy'[sidx]
     return policy.action_map[aidx]
 end
 
-# Method stores record for each evaluated epoch to FiniteHorizonPolicy and returns it
-function addepochrecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy)    
-    global fhepoch
-    fhpolicy.qmat[fhepoch, :, :] = qmat
-    fhpolicy.util[fhepoch, :] = util
-    fhpolicy.policy[fhepoch, :] = policy
+"""
+    addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)
+
+Store record for given stage results to `FiniteHorizonPolicy` and return updated version.
+"""
+function addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)    
+    fhpolicy.qmat[stage, :, :] = qmat
+    fhpolicy.util[stage, :] = util
+    fhpolicy.policy[stage, :] = policy
     return fhpolicy
 end
 
+# MDP given horizon 5 assumes that agent can move 5 times
+function solve(w::FHWrapper; verbose::Bool=false, new_VI::Bool=true)
```

Comment:
As you suggested in #14, I think this should be

```suggestion
function POMDPs.solve(solver::FiniteHorizonSolver, m::MDP)
```